src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs

```diff
@@ -105,42 +114,98 @@ internal static partial class MetadataUtilities
                 modulesByName[name] = modulesByName.ContainsKey(name) ? null : metadata;
             }
 
+            // We don't walk the references of winmd assemblies currently. (That would require
+            // walking references from the current module and also the winmd assemblies.)
+            // So if there are any winmd assemblies, we'll use all assemblies.
+            if (kind == MakeAssemblyReferencesKind.AllReferences && runtimeWinMdBuilder.Any())
+            {
+                kind = MakeAssemblyReferencesKind.AllAssemblies;
+            }
+
             // Build assembly references from modules in primary module manifests.
             var referencesBuilder = ArrayBuilder<MetadataReference>.GetInstance();
-            var identitiesBuilder = (identityComparer == null) ? null : ArrayBuilder<AssemblyIdentity>.GetInstance();
+            var identitiesBuilder = (kind == MakeAssemblyReferencesKind.DirectReferencesOnly) ? ArrayBuilder<AssemblyIdentity>.GetInstance() : null;
+            referencesByIdentity = (kind == MakeAssemblyReferencesKind.AllReferences) ? new Dictionary<string, ImmutableArray<(AssemblyIdentity, MetadataReference)>>(StringComparer.OrdinalIgnoreCase) : null;
+            ModuleMetadata targetModule = null;
             AssemblyIdentity intrinsicsAssembly = null;
+            MetadataReference targetReference = null;
 
             foreach (var metadata in metadataBuilder)
             {
                 if (!IsPrimaryModule(metadata))
                 {
                     continue;
                 }
+
+                var reference = MakeAssemblyMetadata(metadata, modulesByName);
+
+                if (kind == MakeAssemblyReferencesKind.AllReferences)
+                {
+                    var reader = metadata.MetadataReader;
+                    var identity = reader.ReadAssemblyIdentityOrThrow();
+                    if (referencesByIdentity != null)
+                    {
+                        string identityKey = identity.Name;
```

Comment:
I'd inline this variable

Reply:
It's used a couple of times. Since it's not obvious what the dictionary key is, I'll keep the variable.

---
In reply to: [181511479](https://github.com/dotnet/roslyn/pull/25955#discussion_r181511479) [](ancestors = 181511479)

Reply:
I think this would read better: `referencesBySimpleName.TryGetValue(identity.Name, out references)`